File: 3.x.x/crates-utopia/cdn/src/cache/adapter/fastly.rs
//! PHP `Utopia\Cdn\Cache\Adapter\Fastly`.

use std::sync::Arc;

use serde_json::{json, Value};

use super::Adapter;
use crate::http::{default_client, push_percent, send, HttpClient, RequestResult};
use crate::{CdnError, Domain, UnsupportedOperation};

/// Fastly cache-purge adapter.
pub struct Fastly {
    api_token: String,
    domain_key_prefix: String,
    service_id: Option<String>,
    soft_purge: bool,
    api_base: String,
    client: Arc<dyn HttpClient>,
}

impl std::fmt::Debug for Fastly {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Fastly")
            .field("domain_key_prefix", &self.domain_key_prefix)
            .field("service_id", &self.service_id)
            .field("soft_purge", &self.soft_purge)
            .field("api_base", &self.api_base)
            .finish_non_exhaustive()
    }
}

impl Fastly {
    /// Fastly's documented ceiling for one batch surrogate-key purge.
    pub const KEYS_PER_PURGE: usize = 256;

    #[must_use]
    pub fn new(api_token: impl Into<String>, domain_key_prefix: impl Into<String>) -> Self {
        Self {
            api_token: api_token.into(),
            domain_key_prefix: domain_key_prefix.into(),
            service_id: None,
            soft_purge: false,
            api_base: "https://api.fastly.com".into(),
            client: Arc::new(default_client()),
        }
    }

    #[must_use]
    pub fn with_service_id(mut self, service_id: impl Into<String>) -> Self {
        let id = service_id.into();
        self.service_id = if id.is_empty() { None } else { Some(id) };
        self
    }

    #[must_use]
    pub fn with_soft_purge(mut self, soft_purge: bool) -> Self {
        self.soft_purge = soft_purge;
        self
    }

    #[must_use]
    pub fn with_client(mut self, client: Arc<dyn HttpClient>) -> Self {
        self.client = client;
        self
    }

    #[must_use]
    pub fn with_api_base(mut self, api_base: impl Into<String>) -> Self {
        self.api_base = api_base.into();
        self
    }

    fn require_service_id(&self, operation: &str) -> Result<&str, CdnError> {
        match self.service_id.as_deref() {
            Some(id) if !id.is_empty() => Ok(id),
            _ => Err(UnsupportedOperation(format!(
                "Fastly service ID is required for {operation}."
            ))
            .into()),
        }
    }

    /// PHP `encodePath`: percent-encode bytes outside Fastly's allowed set.
    fn encode_path(path: &str) -> String {
        let mut out = String::new();
        for ch in path.chars() {
            if matches!(
                ch,
                'A'..='Z'
                    | 'a'..='z'
                    | '0'..='9'
                    | '-'
                    | '.'
                    | '_'
                    | '~'
                    | '/'
                    | '%'
                    | '?'
                    | '='
                    | '&'
                    | ':'
                    | '+'
            ) {
                out.push(ch);
            } else {
                for byte in ch.encode_utf8(&mut [0; 4]).bytes() {
                    push_percent(&mut out, byte);
                }
            }
        }
        out
    }

    fn send_req(&self, method: &str, path: &str, body: Option<Value>) -> Result<(), CdnError> {
        let result = self.request(method, path, body.as_ref());
        if result.status < 200 || result.status >= 300 {
            return Err(CdnError::runtime(format_error(&result)));
        }
        Ok(())
    }

    fn request(&self, method: &str, path: &str, body: Option<&Value>) -> RequestResult {
        let mut headers = vec![
            ("User-Agent", "Utopia CDN Fastly Adapter".to_owned()),
            ("Fastly-Key", self.api_token.clone()),
            ("Accept", "application/json".to_owned()),
        ];
        if self.soft_purge {
            headers.push(("Fastly-Soft-Purge", "1".to_owned()));
        }
        if body.is_some() {
            headers.push(("Content-Type", "application/json".to_owned()));
        }
        send(
            self.client.as_ref(),
            method,
            &format!("{}{path}", self.api_base),
            &headers,
            body,
        )
    }
}

fn format_error(result: &RequestResult) -> String {
    let message = result
        .error
        .clone()
        .or_else(|| {
            result
                .response
                .get("msg")
                .or_else(|| result.response.get("detail"))
                .and_then(Value::as_str)
                .map(str::to_owned)
        })
        .unwrap_or_else(|| "Unknown purge error".into());
    format!(
        "Fastly purge failed with status {}: {message}",
        result.status
    )
}

impl Adapter for Fastly {
    fn purge_paths(&self, domain: &str, paths: &[String]) -> Result<(), CdnError> {
        let domain = Domain::validate(domain)?;
        let paths = Domain::validate_paths(paths)?;
        if paths.is_empty() {
            return Ok(());
        }
        for path in paths {
            self.send_req(
                "POST",
                &format!("/purge/{domain}{}", Self::encode_path(&path)),
                None,
            )?;
        }
        Ok(())
    }

    fn purge_domain(&self, domain: &str) -> Result<(), CdnError> {
        let domain = Domain::validate(domain)?;
        self.purge_keys(&[format!("{}{domain}", self.domain_key_prefix)])
    }

    fn purge_keys(&self, keys: &[String]) -> Result<(), CdnError> {
        if keys.is_empty() {
            return Ok(());
        }
        let service = self.require_service_id("cache key purging")?;
        for chunk in keys.chunks(Self::KEYS_PER_PURGE) {
            self.send_req(
                "POST",
                &format!("/service/{service}/purge"),
                Some(json!({ "surrogate_keys": chunk })),
            )?;
        }
        Ok(())
    }

    fn purge_zone(&self) -> Result<(), CdnError> {
        let service = self.require_service_id("zone purging")?;
        self.send_req("POST", &format!("/service/{service}/purge_all"), None)
    }
}

File: src/Cdn/Cache/Adapter/Fastly.php
<?php

namespace Utopia\Cdn\Cache\Adapter;

use Psr\Http\Client\ClientExceptionInterface;
use Psr\Http\Client\ClientInterface;
use Utopia\Cdn\Cache\Adapter;
use Utopia\Cdn\Domain;
use Utopia\Cdn\Exception\UnsupportedOperation;
use Utopia\Client;
use Utopia\Client\Adapter\Curl\Client as CurlAdapter;
use Utopia\Psr7\Header;
use Utopia\Psr7\Method;
use Utopia\Psr7\Request\Factory as RequestFactory;

class Fastly implements Adapter
{
    /**
     * Fastly's documented ceiling for one batch surrogate key purge.
     */
    public const int KEYS_PER_PURGE = 256;

    private readonly ClientInterface $client;

    /**
     * Fastly cannot purge by host: its purge API offers URL, surrogate key and whole-service purges
     * and nothing in between. A domain is therefore addressed by the surrogate key the origin
     * attaches to every response it serves for that domain, and this adapter has to be told how
     * those keys are named — hence a required prefix rather than an optional one.
     *
     * @param string $domainKeyPrefix Prefix of the per-domain surrogate key. Pass '' when the key is the bare hostname.
     */
    public function __construct(
        private readonly string $apiToken,
        private readonly string $domainKeyPrefix,
        private readonly ?string $serviceId = null,
        private readonly bool $softPurge = false,
        ?ClientInterface $client = null,
        private readonly string $apiBase = 'https://api.fastly.com',
    ) {
        $this->client = $client ?? new Client(new CurlAdapter());
    }

    public function purgePaths(string $domain, array $paths): void
    {
        $domain = Domain::validate($domain);
        $paths = Domain::validatePaths($paths);

        if ($paths === []) {
            return;
        }

        // A URL purge carries one URL, so there is nothing to batch.
        foreach ($paths as $path) {
            $this->send(Method::POST, '/purge/' . $domain . $this->encodePath($path));
        }
    }

    /**
     * Purges the domain's surrogate key, leaving every other domain on the service cached.
     */
    public function purgeDomain(string $domain): void
    {
        $this->purgeKeys([$this->domainKeyPrefix . Domain::validate($domain)]);
    }

    public function purgeKeys(array $keys): void
    {
        if ($keys === []) {
            return;
        }

        $this->requireServiceId('cache key purging');

        // Keys travel in the request body rather than the URL, so they are sent as
        // given: percent-encoding one would purge a key the origin never attached.
        foreach (array_chunk($keys, self::KEYS_PER_PURGE) as $chunk) {
            $this->send(Method::POST, '/service/' . $this->serviceId . '/purge', ['surrogate_keys' => $chunk]);
        }
    }

    /**
     * Purges every object on the service, whatever domain it belongs to.
     *
     * Fastly documents purge_all as taking up to two minutes, being incompatible with soft purge,
     * and likely to spike origin traffic on a busy service. Prefer a surrogate key purge.
     */
    public function purgeZone(): void
    {
        $this->requireServiceId('zone purging');

        $this->send(Method::POST, '/service/' . $this->serviceId . '/purge_all');
    }

    /**
     * Reported as an unsupported operation rather than a failure: a token without a service ID can
     * still purge URLs, so this is a gap in what the adapter was configured for and not a purge that
     * was attempted and went wrong. A caller can tell the two apart and decide whether to carry on.
     */
    private function requireServiceId(string $operation): void
    {
        if ($this->serviceId === null || $this->serviceId === '') {
            throw new UnsupportedOperation('Fastly service ID is required for ' . $operation . '.');
        }
    }

    private function encodePath(string $path): string
    {
        return (string) preg_replace_callback(
            '/[^A-Za-z0-9\-._~\/%?=&:+]/u',
            static fn(array $match): string => rawurlencode($match[0]),
            $path,
        );
    }

    /**
     * @param array<string, mixed>|null $body
     */
    private function send(string $method, string $url, ?array $body = null): void
    {
        $result = $this->request($method, $url, $body);

        if ($result['statusCode'] < 200 || $result['statusCode'] >= 300) {
            throw new \RuntimeException($this->formatError($result));
        }
    }

    /**
     * @param array{statusCode:int,response:array<string, mixed>|string|null,error:string|null} $result
     */
    private function formatError(array $result): string
    {
        $message = $result['error'] ?? null;

        if (\is_array($result['response'])) {
            $message ??= $result['response']['msg'] ?? $result['response']['detail'] ?? null;
        }

        $message ??= 'Unknown purge error';

        return 'Fastly purge failed with status ' . $result['statusCode'] . ': ' . $message;
    }

    /**
     * @param array<string, mixed>|null $body
     * @return array{statusCode:int,response:array<string, mixed>|string|null,error:string|null}
     */
    private function request(string $method, string $url, ?array $body = null): array
    {
        $factory = new RequestFactory();
        $request = $body === null
            ? $factory->createRequest($method, $this->apiBase . $url)
            : $factory->json($method, $this->apiBase . $url, $body);

        $request = $request
            ->withHeader(Header::USER_AGENT, 'Utopia CDN Fastly Adapter')
            ->withHeader('Fastly-Key', $this->apiToken)
            ->withHeader(Header::ACCEPT, 'application/json');

        if ($this->softPurge) {
            $request = $request->withHeader('Fastly-Soft-Purge', '1');
        }

        try {
            $response = $this->client->sendRequest($request);
        } catch (ClientExceptionInterface $error) {
            return ['statusCode' => 0, 'response' => null, 'error' => $error->getMessage()];
        }

        $contents = (string) $response->getBody();

        try {
            $decoded = json_decode($contents, true, flags: JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            $decoded = $contents;
        }

        return ['statusCode' => $response->getStatusCode(), 'response' => $decoded, 'error' => null];
    }
}
